CMakeLists.txt

```diff
@@ -275,7 +275,8 @@ if (WIN32)
 endif ()
 
 # Include directories for the previous dependencies.
-set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${ARMADILLO_INCLUDE_DIRS})
+set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${ARMADILLO_INCLUDE_DIRS}
+    ${CMAKE_SOURCE_DIR}/src/mlpack/core/optimizers/ensmallen)
```

Comment:
Do you think it might be possible to use CMake to check if ensmallen is available on the target system and then, e.g., use a git submodule or download and unpack the code and something like this if it isn't available?  I can probably write some code to do this in CMake, the trick is just figuring out the right way to do it.

Keeping ensmallen in the repo like this is a bit of a maintenance burden that ideally I'd like to avoid. :)

We can probably expect ensmallen to become readily available in different package managers soon, but it will probably take a while.

Reply:
The main issue with git submodule or a download step, is that you can't just download mlpack, and build it without an internet connection. Maybe the case doesn't really exist in the real world.

Reply:
That's true.  Over time I think we can expect that people have ensmallen installed, but at the moment it would be another dependency that might not be available in people's package managers.  So maybe we could do the following:

 - For a release, package `ensmallen-x.y.z.tar.gz` (the newest version) in `deps/`.  CMake can try to find ensmallen on the system but if it is not available then it will unpack and use the one in `deps/`.
 - For a git repository, CMake can try to download the newest version of ensmallen if it isn't available from the website.

What do you think, would that be more reasonable?

Reply:
That sounds like a good idea to me.

>  I can probably write some code to do this in CMake, the trick is just figuring out the right way to do it.

If you have the time, I guess we could just open a new PR and merge this one after the CMake modification?

Reply:
Sure, I think that I can write this CMake script by tomorrow.

Reply:
Here is an attempt:

https://gist.github.com/acd87095d44d73d1355a81a89dd9b87d 

It is not fully debugged but I am about to be on a really long flight with no internet so I wanted to get it uploaded. :) Let me know if there are problems and I will try to debug when I am able.